aten/src/ATen/native/cuda/CuFFTPlanCache.h

```diff
@@ -15,51 +15,102 @@
+  // These include additional batch dimension as well.
+  int64_t sizes_[max_rank + 1];
+  int64_t input_strides_[max_rank + 1];
+  int64_t output_strides_[max_rank + 1];
+  CuFFTTransformType fft_type_;
+  ScalarType value_type_;
+
+  CuFFTParams() = default;
+
+  CuFFTParams(IntArrayRef in_strides, IntArrayRef out_strides,
+      IntArrayRef signal_sizes, CuFFTTransformType fft_type, ScalarType value_type) {
+    // Padding bits must be zeroed for hashing
+    memset(this, 0, sizeof(*this));
+    signal_ndim_ = signal_sizes.size() - 1;
+    fft_type_ = fft_type;
+    value_type_ = value_type;
+
+    TORCH_INTERNAL_ASSERT(in_strides.size() == signal_sizes.size());
+    TORCH_INTERNAL_ASSERT(out_strides.size() == signal_sizes.size());
+    TORCH_INTERNAL_ASSERT(1 <= signal_ndim_ && signal_ndim_ <= max_rank);
+
+    std::copy(signal_sizes.begin(), signal_sizes.end(), sizes_);
+    std::copy(in_strides.begin(), in_strides.end(), input_strides_);
+    std::copy(out_strides.begin(), out_strides.end(), output_strides_);
+  }
 };
 
-// NB: This can't be a constructor, because then CuFFTParams
-// would not be a POD anymore.
-static inline void setCuFFTParams(CuFFTParams* params,
-    const Tensor& input, int64_t signal_ndim, bool complex_input,
-    bool complex_output, IntArrayRef checked_signal_sizes, bool onesided) {
-
-  memset(params, 0, sizeof(CuFFTParams));
-  params->scalar_type_ = input.scalar_type();
-  for (int i = 0; i != input.dim(); ++i) {
-    params->input_sizes_[i] = input.size(i);
-    if (input.size(i) != 1) {
-      params->input_strides_[i] = input.stride(i);
+static_assert(std::is_trivial<CuFFTParams>::value, "");
+
+inline bool cufft_complex_input(CuFFTTransformType type) {
+  switch (type) {
+    case CuFFTTransformType::C2C:
+    case CuFFTTransformType::C2R: {
+      return true;
+    }
+
+    case CuFFTTransformType::R2C: {
+      return false;
+    }
+  }
+  TORCH_INTERNAL_ASSERT(false);
+}
+
+inline bool cufft_complex_output(CuFFTTransformType type) {
+  switch (type) {
+    case CuFFTTransformType::C2C:
+    case CuFFTTransformType::R2C: {
+      return true;
+    }
+
+    case CuFFTTransformType::C2R: {
+      return false;
     }
   }
-  params->signal_ndim_ = (uint8_t) signal_ndim;
-  params->complex_input_ = complex_input;
-  params->complex_output_ = complex_output;
-  for (size_t i = 0; i != checked_signal_sizes.size(); ++i) {
-    params->signal_sizes_[i] = checked_signal_sizes[i];
+  TORCH_INTERNAL_ASSERT(false);
+}
+
+inline CuFFTTransformType GetCuFFTTransformType(bool complex_input, bool complex_output) {
+  if (complex_input && complex_output) {
+    return CuFFTTransformType::C2C;
+  } else if (complex_input && !complex_output) {
+    return CuFFTTransformType::C2R;
+  } else if (!complex_input && complex_output) {
+    return CuFFTTransformType::R2C;
   }
-  params->onesided_ = onesided;
+  TORCH_CHECK(false, "Real to real FFTs are not supported");
```

Comment:
If this error only occurs due to an internal logic failure (that is, not due to user actions), then make it a TORCH_INTERNAL_ASSERT